Fixes Cigarettes being lit in no-oxygen environments

timothymtorres opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
This closes #26487 which has been a known bug for years.  Everyone knows how **EXTREMELY CRITICAL** cigarettes are to game balance.   Plasmamen being able to smoke in a room with _no oxygen_?  Spessmen smoking in the _hard vacuum of space_?  What is this madness?!?  

I must put an end to these OP tactics.

My changes make it so that if a cigarette has either one mole of oxygen in the current atmosphere or is injected with oxygen reagents, then it can be smoked. 

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->
Realism for the most part.  I thought that issue was hilarious and wanted to add it just for the meme.  I did make it so that "Space" cigarettes can be smoked in space by changing the default starting reagents to include oxygen.  The description was also updated to hint that they can be smoked anywhere. 

Also I added a `has_gas` proc to the turf class.  I have seen a few places in the code where this can be used and once this gets merged, I will be doing a refactor to clean some code involving it.  (bonfires, plasmamen code, etc.)  

## Changelog
:cl:
fix: Fixed cigarettes being lit in no-oxygen environments
add: Cigarettes can be injected with oxygen to allow smoking anywhere
code: The cigarette brand 'Space cigarettes' now has oxygen in it's default reagents (allowing it to be smoked in space)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

timothymtorres committed:
Add has_gas proc for open turfs

timothymtorres committed:
Fix cigarettes burning without oxygen present

timothymtorres committed:
Add description for cigarettes without oxygen

timothymtorres committed:
Add lit cigarettes extinguishing without oxygen

timothymtorres committed:
Change space_cigarette so it can be smoked in space

timothymtorres committed:
Change space cigarette descriptions

Jared-Fogle commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -157,12 +157,16 @@ CIGARETTE PACKETS ARE IN FANCY.DM
+	if(lit || smoketime <= 0)
 		return ..()
+	if(!reagents.has_reagent(/datum/reagent/oxygen)) //cigarettes need oxygen
+		var/turf/open/O = get_turf(src)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

timothymtorres replied:
Fixed with 2207989

Jared-Fogle commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -265,13 +268,17 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 
 /obj/item/clothing/mask/cigarette/process(delta_time)
-	var/turf/location = get_turf(src)
+	var/turf/open/O = get_turf(src)
```
https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

Jared-Fogle replied:
You *removed* the good name from it 🥴

timothymtorres replied:
Fixed with 2207989.  Eh, I think having two variables named loc and location is bad.  The O was taken from another part of the code I was looking at.

Jared-Fogle commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -92,6 +92,13 @@
 /turf/open/return_analyzable_air()
 	return return_air()
 
+/turf/open/proc/has_gas(datum/gas, mole_amount)
```
Document this.

Also, this feels like it should be a macro instead.

timothymtorres replied:
Moved this to `datum/gas_mixture`

timothymtorres committed:
Refactor variable name for cigarette location

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -92,6 +92,13 @@
 /turf/open/return_analyzable_air()
 	return return_air()
 
+/turf/open/proc/has_gas(datum/gas, mole_amount)
```
1: Why is this on /turf/open instead of /datum/gas_mixture?
2: The gas argument should not be typed as a datum, it's a typepath.

timothymtorres replied:
So my thought process went like this:

I need a way to check if a turf has a certain moles of gas.  TURF -> GASES -> GAS
Thus the root class that I thought I needed was Turf.  It appears /datum/gas_mixture is a global class used for all things.  It might be useful to have the proc inside that class instead.  Let me tinker with it since I'm still pretty new to coding in DM.

timothymtorres replied:
Both of these are resolved with 73324af

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -92,6 +92,13 @@
 	return return_air()
 
+/turf/open/proc/has_gas(datum/gas, mole_amount)
+	if(air && air.total_moles())
```
total_moles() is not relevant here

timothymtorres replied:
So the reason for using `total_moles()` here is due to the possibility of space turfs.  They have an air variable but it creates a runtime since it's different from regular turfs.  The total moles is a cheap hack to check and see if there is any air pressure.

I used this as my reference. 

https://github.com/tgstation/tgstation/blob/2a90fdd75e1488c4be03cdfcce20b70de35a173d/code/modules/mob/living/carbon/human/species_types/plasmamen.dm#L49

LemonInTheDark replied:
oh for god's sake you're right

timothymtorres replied:
73324af avoids using `total_moles()` since the `has_gas` proc uses `ASSERT_GAS` which creates a temporary gas on the space turf to check if there is gas.  Then later it gets deleted by the GC

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -92,6 +92,13 @@
 
+/turf/open/proc/has_gas(datum/gas, mole_amount)
+	if(air && air.total_moles())
+		var/loc_gases = air.gases
+		if(loc_gases[gas] && loc_gases[gas][MOLES] >= mole_amount)
+			return TRUE
```
I'd prefer ASSERT_GAS to be used here, to maintain consistency with other uses of gaslists.

timothymtorres replied:
This was added to the `has_gas` proc.

LemonInTheDark reviewed (changes requested):
As I've said on discord, my gut says that we shouldn't use procs to wrap around gaslist operations, as it will lead to misuse of the framework. That aside, I have some concerns about how you're doing this.

LemonInTheDark commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -157,12 +157,16 @@ CIGARETTE PACKETS ARE IN FANCY.DM
+	if(lit || smoketime <= 0)
 		return ..()
+	if(!reagents.has_reagent(/datum/reagent/oxygen)) //cigarettes need oxygen
+		var/turf/open/location = get_turf(src)
```
You really shouldn't be using the turf for these checks, for reasons laid out in the original issue. Check to see if loc isn't null, and use loc.return_air(). Or hell, call return_air() normally, it'll do the check for you.

timothymtorres replied:
Initially when I first coded this, I tried to use loc to get the item location but it only worked if it was on the ground.  If the item was in a container or on the players mask slot, it would give me the other object instead of the turf.  This is why I used `get_turf`.  I will try it again with `return_air` and see if it will be any different.

LemonInTheDark replied:
To be clear, this would only work if you put has_gas on gasmixtures rather then turfs

timothymtorres replied:
Resolved with 73324af

Also the `return_air` suggestion was very helpful.  Thanks!

timothymtorres commented:
> As I've said on discord, my gut says that we shouldn't use procs to wrap around gaslist operations, as it will lead to misuse of the framework. That aside, I have some concerns about how you're doing this.

There are several files where custom `has_gas` procs were used or the mole variable was directly referenced due to not having access to the operation.

See the following:

https://github.com/tgstation/tgstation/blob/2a90fdd75e1488c4be03cdfcce20b70de35a173d/code/modules/mob/living/carbon/human/species_types/plasmamen.dm#L54

https://github.com/tgstation/tgstation/blob/849692d113e37a0599ad8eaaef6916992e9f61c9/code/modules/hydroponics/grown/towercap.dm#L230

https://github.com/tgstation/tgstation/blob/7976305e71faa937d4c1b4e312501599a2720885/code/modules/mob/living/life.dm#L109

coiax commented:
Glad to see one of my classic issues being dealt with.

Don't forget you need CO2 emissions by the combusting cigarette.

Pugie commented:
smoking in space is badass

timothymtorres commented:
> smoking in space is badass
> 
> Sure realism is cool and all but this is a sci-fi cartoonish game that doesn't take itself all that seriously this pr subtracts more than it adds imo.

Smoking in space is not removed.  Smoking without *oxygen* however will no longer be possible.

While these both may seem to be identical at first, there is a distinct difference.  Injecting any kind of cigarette with oxygen will result in it being able to smoke in *any environment*.  Additionally, the cigarette brand **"Space Cigarettes"** has had its starting reagents changed to include oxygen so that they may be smoked in space as a pun.

Pugie commented:
AWESOME AND BADASS MCSIR

timothymtorres committed:
Move has_gas proc to /datum/gas_mixture

timothymtorres committed:
Refactor return_air procs for cigarettes

Also changed the reagent amounts and time for space cigarettes.

timothymtorres committed:
Fix undefinied location var for cigarettes

Mothblocks commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -106,6 +106,12 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	var/cached_gases = gases
 	TOTAL_MOLES(cached_gases, .)
 
+	/// Checks to see if gas amount exists in mixture
```
```suggestion
/// Checks to see if gas amount exists in mixture
```

timothymtorres replied:
All of the documentation for the procs have 1 tab in front of them.

Cyberboss replied:
Yes, and it's wrong

LemonInTheDark replied:
Dennok did me dirty when he documented these

timothymtorres replied:
I see.  Well it won't be much hassle to remove all the tabs from the file for the documentation.

timothymtorres replied:
228b21d removes all the whitespace from the file.

timothymtorres committed:
Change cigarette butt to spawn on turf

timothymtorres committed:
Add warning comment to has_gas proc

LemonInTheDark commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -157,12 +157,16 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 		return ..()
+	if(!reagents.has_reagent(/datum/reagent/oxygen)) //cigarettes need oxygen
+		var/datum/gas_mixture/air = return_air()
+		if(!air.has_gas(/datum/gas/oxygen, 1)) //or oxygen on a tile to burn
```
It's rare, but return_air can in some cases return a null value. You should check for it here.

timothymtorres replied:
That is some strange behavior...  I wonder if it's caused by the tile z-level being completely removed.  (like if a bomb goes off on ice and you can fall through the floor, etc.)

Regardless, it has been addressed with d85e3b4

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -106,6 +106,13 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	TOTAL_MOLES(cached_gases, .)
 
+	/// Checks to see if gas amount exists in mixture
+	/// WARNING - DO NOT USE THIS IN PERFORMANCE INTENSIVE CODE!!!
```
Explain why

timothymtorres replied:
Code comment added with 942c867

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -106,6 +106,13 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 
+	/// Checks to see if gas amount exists in mixture
+	/// WARNING - DO NOT USE THIS IN PERFORMANCE INTENSIVE CODE!!!
+/datum/gas_mixture/proc/has_gas(gas_id, amount=0)
+	ASSERT_GAS(gas_id, src)
+	var/list/cached_gases = gases
+	return amount < cached_gases[gas_id][MOLES]
```
For one access caching isn't worth it. Store the result of the check as a var, and call garbage_collect() before returning so gaslists don't get a lot of empty entries.

timothymtorres replied:
942c867 does all the above.  Also good to know that `garbage_collect` needs to be called whenever `ASSERT_GAS` is used.

timothymtorres committed:
Remove whitespace from code documentation

timothymtorres committed:
Add check to see if air is null

In some rare cases return_air() can return null so we need to prepare for this.

timothymtorres committed:
Fix logical operator to not use bitflag OR

LemonInTheDark commented on code/game/objects/items/cigs_lighters.dm:
```diff
@@ -269,6 +272,11 @@ CIGARETTE PACKETS ARE IN FANCY.DM
 		M.IgniteMob()
+	if(!reagents.has_reagent(/datum/reagent/oxygen)) //cigarettes need oxygen
+		var/datum/gas_mixture/air = return_air()
+		if(air || !air.has_gas(/datum/gas/oxygen, 1)) //or oxygen on a tile to burn
```
Missed a ! here

timothymtorres replied:
Good catch!  Fixed with 7c6eb3a

timothymtorres committed:
Refactor has_gas proc to use garbage_collect()

timothymtorres committed:
Fix any air causing cigarettes to extinguish itself

Whoops!  Forgot to include a ! here.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -101,12 +101,21 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
-	/// Calculate pressure in kilopascals
+/// Checks to see if gas amount exists in mixture.
+/// Do NOT use this in code where performance matters!
+/// It's faster to just manually check the gas with a direct var lookup
```
```suggestion
/// It's better to batch calls to garbage_collect(), especially in places where you're checking many gastypes
```

timothymtorres replied:
Done

timothymtorres committed:
Update code/modules/atmospherics/gasmixtures/gas_mixture.dm

Co-authored-by: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -101,12 +101,21 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
 	TOTAL_MOLES(cached_gases, .)
 
-	/// Calculate pressure in kilopascals
+/// Checks to see if gas amount exists in mixture.
+/// Do NOT use this in code where performance matters!
+/// It's better to batch calls to garbage_collect(), especially in places where you're checking many gastypes
```
Sorry for not mentioning this earlier, but this should be in the [dm doc format](https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md), so it's parsed properly <

Mothblocks replied:
This works, actually. Up to user preference.

LemonInTheDark replied:
Last time I saw it tried shit got messed up, but I could be misremembering

LemonInTheDark replied:
Brain no work so good, code document well

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/gas_mixture.dm:
```diff
@@ -49,49 +49,49 @@ GLOBAL_LIST_INIT(gaslist_cache, init_gaslist_cache())
-	///assert_gas(gas_id) - used to guarantee that the gas list for this id exists in gas_mixture.gases.
-	//Must be used before adding to a gas. May be used before reading from a gas.
+///assert_gas(gas_id) - used to guarantee that the gas list for this id exists in gas_mixture.gases.
+//Must be used before adding to a gas. May be used before reading from a gas.
```
```suggestion
///Must be used before adding to a gas. May be used before reading from a gas.
```
Just gonna webedit this in don't mind meeee

LemonInTheDark committed:
Fixes an improperly documenting comment

tgstation-server commented:
Your Fix/Feature pull request delta is now above zero (0). Feel free to make Feature/Tweak/Balance PRs.

coiax commented:
Finally, my stealth nerf to plasmamen smoking in space has been completed.
